File: src/main/java/org/taktik/icure/logic/impl/TimeTableLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.TimeTableDAO;
import org.taktik.icure.entities.TimeTable;
import org.taktik.icure.exceptions.DeletionException;
import org.taktik.icure.logic.TimeTableLogic;
import org.taktik.icure.logic.ICureSessionLogic;

import java.util.List;

@Service
public class TimeTableLogicImpl extends GenericLogicImpl<TimeTable, TimeTableDAO> implements TimeTableLogic {

	private TimeTableDAO timeTableDAO;
	private ICureSessionLogic sessionLogic;

	@Override
	public TimeTable createTimeTable(TimeTable timeTable) {
		return timeTableDAO.create(timeTable);
	}

    @Override
    public List<String> deleteTimeTables(List<String> ids) throws DeletionException {
        try {
            deleteEntities(ids);
            return ids;
        } catch (Exception e) {
            throw new DeletionException(e.getMessage(), e);
        }
    }

    @Override
	public TimeTable getTimeTable(String timeTableId) {
		return timeTableDAO.get(timeTableId);
	}

	@Override
	public List<TimeTable> getTimeTablesByPeriodAndAgendaId(Long startDate, Long endDate, String agendaId){
		return timeTableDAO.listTimeTableByPeriodAndAgendaId(startDate,endDate,agendaId);
	}

	@Override
	public List<TimeTable> getTimeTablesByAgendaId(String agendaId){
		return timeTableDAO.listTimeTableByAgendaId(agendaId);
	}
	@Override
	public TimeTable modifyTimeTable(TimeTable timeTable) {
		return timeTableDAO.save(timeTable);
	}

	@Autowired
  public void setTimeTableDAO(TimeTableDAO timeTableDAO) {
		this.timeTableDAO = timeTableDAO;
	}

	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	protected TimeTableDAO getGenericDAO() {
		return timeTableDAO;
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/TimeTableLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.icure.asyncdao.TimeTableDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.TimeTableLogic
import org.taktik.icure.entities.TimeTable

@Service
class TimeTableLogicImpl(private val timeTableDAO: TimeTableDAO, private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<TimeTable, TimeTableDAO>(sessionLogic), TimeTableLogic {
	override suspend fun createTimeTable(timeTable: TimeTable) = fix(timeTable) { timeTable ->
		timeTableDAO.create(timeTable)
	}

	override fun deleteTimeTables(ids: List<String>): Flow<DocIdentifier> {
		return deleteEntities(ids)
	}

	override suspend fun getTimeTable(timeTableId: String): TimeTable? {
		return timeTableDAO.get(timeTableId)
	}

	override fun getTimeTablesByPeriodAndAgendaId(startDate: Long, endDate: Long, agendaId: String): Flow<TimeTable> = flow {
		emitAll(timeTableDAO.listTimeTableByPeriodAndAgendaId(startDate, endDate, agendaId))
	}

	override fun getTimeTablesByAgendaId(agendaId: String): Flow<TimeTable> = flow {
		emitAll(timeTableDAO.listTimeTableByAgendaId(agendaId))
	}

	override suspend fun modifyTimeTable(timeTable: TimeTable) = fix(timeTable) { timeTable ->
		timeTableDAO.save(timeTable)
	}

	override fun getGenericDAO(): TimeTableDAO {
		return timeTableDAO
	}
}
